### DISCUSSION TITLE
bind_text within for loop

### DISCUSSION THREAD

**Discussion by laserir** [zauberzeug/nicegui#1561] [Q&A]
### Question

Hi folks,
I tried to use the following code, to create some `ui.card()` with `ui.badge` each. This doesnt work (TypeError: list indices must be integers or slices, not dict)...
```py
for index, system in enumerate(o.data["system"]):
    with ui.card().tight():
        ui.badge('0', color='red').props('floating').bind_text_from(o, 'data', backward=lambda index=index: (len(o.data["system"][index]["error"])))
```
...but this works BUT attaches the value of the last entry to all badges:
```py
ui.badge('0', color='red').props('floating').bind_text_from(o, 'data', backward=lambda _: len(o.data["system"][index]["error"]))
```
Where am I thinking wrong? :(

**Comment by laserir** [comment]
Solved! If anyone is running in a similar situation... see below.
```py
ui.badge('0', color='red').props('floating') \
    .bind_text_from(o, 'data', backward=lambda data=o.data, index=index: len(data["system"][index]["error"]))
    .bind_visibility_from(o, 'data', backward=lambda data=o.data, index=index: len(data["system"][index]["error"]))
```

**Comment by laserir** [comment]
Doesnt seem to update when values change. This is the structure o.status:
`{'system': {'controller': {'errors': [{'entity_id': 'controller', 'error_id': 'err_hardware_serial', 'level': 40, 'timestamp': datetime.datetime(2023, 10, 6, 20, 41, 24, 346107), 'details': None}]}, 'maze': {'errors': [{'entity_id': 'maze', 'error_id': 'test_error', 'level': 30, 'timestamp': datetime.datetime(2023, 10, 6, 20, 41, 35, 715200), 'details': None}]}}`

I have tried to bind the number of errors like:

```py
for system_key, system_data in o.status["system"].items():
    ui.label('').bind_text_from(o, 'status', backward=lambda system_key=system_key: o.get_errors_count("system", system_key))
```
It only updates when reloading the page, but not if o.status changes. Might that be because it is nested? Any ideas how to trigger the binding? (ui.update() wont work as well)

**Reply by falkoschindler** [reply]
Can you, please, provide a minimal reproducible code example? The snippets you provided require quite some effort to be put together and to recreate the problem you're facing. Preparing a short and concise example that can be easily copy-and-pasted into a Python script allows us and the community to help more efficiently. Thanks!

**Comment by laserir** [comment]
Here is a minimal example reproducing the problem. I found out that this only occurs when the bound instance/property is manipulated within the same page function as the binding!
```py
from nicegui import ui
import time

class Test:
    def __init__(self):
        self.status = {
            "system": {
                "test_value": 0
            }
        }

    def increase_test_value(self):
        self.status["system"]["test_value"] += 1
        print(self.status["system"]["test_value"])

    def get_test_value(self):
        return self.status["system"]["test_value"]

@ui.page('/')
def main_page() -> None:
    print(test_instance.get_test_value())  # Output: 0
    ui.label().bind_text_from(test_instance, 'status', backward=lambda _:test_instance.get_test_value())
    test_instance.increase_test_value()    # Output: 1, but label still displays 0

test_instance = Test()
ui.run()
```

**Reply by falkoschindler** [reply]
Hm, I boiled it down to this:
```py
state = {'count': 0}

@ui.page('/')
def main_page() -> None:
    ui.label().bind_text_from(state, 'count')
    state['count'] += 1
    print(state)  # Output: 1, but label still displays 0
```

But I've no idea why the binding is out of sync...

**Reply by laserir** [reply]
Do you want me to open an issue for that?

**Reply by falkoschindler** [reply]
@laserir Thanks, but it's not necessary in this case. We already have created a PR fixing this bug: https://github.com/zauberzeug/nicegui/pull/1775. 🙂
